### PR TITLE
net: Prevent accidental circuit sharing when using Tor stream isolation

### PR DESCRIPTION
**Pull request by laanwj**
Add a class TorsStreamIsolationCredentialsGenerator that generates unique credentials based on a randomly generated session prefix and an atomic counter. Use this in `ConnectThroughProxy` instead of a simple atomic int counter.

This makes sure that different launches of the application won't share the same credentials, and thus circuits, even in edge cases.

Example with `-debug=proxy`:
```
2025-03-31T16:30:27Z [proxy] SOCKS5 sending proxy authentication 0afb2da441f5c105-0:0afb2da441f5c105-0
2025-03-31T16:30:31Z [proxy] SOCKS5 sending proxy authentication 0afb2da441f5c105-1:0afb2da441f5c105-1
```

Thanks to hodlinator in https://github.com/bitcoin/bitcoin/pull/32166#discussion_r2020973352 for the idea.

### CODE REVIEW

**Review by 1440000bytes — COMMENTED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/32176/commits/3ab2cb2c291df90c605c9976351a6bb03401786d

**Review by eval-exec — COMMENTED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/32176/commits/3ab2cb2c291df90c605c9976351a6bb03401786d

File: src/init.cpp [hunk 1]
```diff
@@ -1578,14 +1578,14 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     if (proxyArg != "" && proxyArg != "0") {
         Proxy addrProxy;
         if (IsUnixSocketPath(proxyArg)) {
-            addrProxy = Proxy(proxyArg, proxyRandomize);
+            addrProxy = Proxy(proxyArg, /*tor_stream_isolation=*/proxyRandomize);
```
Comment (eval-exec) [inline src/init.cpp:1581]: How about change `proxyRandomize` to `tor_stream_isolation` too.

Comment (laanwj) [reply src/init.cpp:1581]: i'm not sure. i don't want to go as far as to change interfaces: configuration option name `-proxyrandomize` or the RPC field `proxy_randomize_credential` of `getnetworkinfo`. It would be clearer but is not worth the disruption. And this variable name seems a direct reflection of the option name.

File: src/netbase.cpp [hunk 2]
```diff
@@ -725,6 +725,41 @@ bool IsProxy(const CNetAddr &addr) {
...
+
+    /** Return the next unique proxy credentials. */
+    ProxyCredentials Generate() {
+        ProxyCredentials auth;
+        auth.username = auth.password = strprintf("%s%i", m_prefix, m_counter);
+        ++m_counter;
+        return auth;
+    }
+
+    /* Size of session prefix in bytes. */
```
Comment (laanwj) [inline src/netbase.cpp:747]: Self-nit: should be `/**` comment for doxygen.

File: src/netbase.cpp [hunk 3]
```diff
@@ -725,6 +725,41 @@ bool IsProxy(const CNetAddr &addr) {
     return false;
 }
 
+/**
+ * Generate unique credentials for Tor stream isolation. Tor will create
+ * seperate circuits for socks5 proxy connections with different credentials, which
```
Comment (laanwj) [inline src/netbase.cpp:730]: Self-nit: SOCKS5

**Review by hodlinator — COMMENTED** [review]
Code review 297ddaa430e99ad62e6bbdcc475ede7e04da547f

Remaining inline Q regarding naming + nits.

File: src/init.cpp [hunk 4]
```diff
@@ -1578,14 +1578,14 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
...
         } else {
             const std::optional<CService> proxyAddr{Lookup(proxyArg, 9050, fNameLookup)};
             if (!proxyAddr.has_value()) {
                 return InitError(strprintf(_("Invalid -proxy address or hostname: '%s'"), proxyArg));
             }
 
-            addrProxy = Proxy(proxyAddr.value(), proxyRandomize);
+            addrProxy = Proxy(proxyAddr.value(), /*tor_stream_isolation=*/proxyRandomize);
         }
```
Comment (hodlinator) [inline src/init.cpp:1590]: nit: Could mention edge-cases from https://github.com/bitcoin/bitcoin/pull/32166#discussion_r2021044241 in PR description:
```diff
- This makes sure that different launches of the application won't share the same credentials, and thus circuits, even in edge cases.
+ This makes sure that different launches of the application won't share the same credentials, and thus circuits, even in edge cases like:
+ * restarting bitcoind in quick succession, with counter always resetting to 0
+ * multiple bitcoinds running in parallel
```
(Uses "bitcoind" over "daemon" as the former is more specific, avoids any confusion with Tor daemon).

File: src/netbase.h [hunk 5]
```diff
@@ -58,14 +58,14 @@ bool IsUnixSocketPath(const std::string& name);
...
-    explicit Proxy(const std::string path, bool _randomize_credentials = false) : m_unix_socket_path(path), m_is_unix_socket(true), m_randomize_credentials(_randomize_credentials) {}
+    Proxy() : m_is_unix_socket(false), m_tor_stream_isolation(false) {}
+    explicit Proxy(const CService& _proxy, bool tor_stream_isolation = false) : proxy(_proxy), m_is_unix_socket(false), m_tor_stream_isolation(tor_stream_isolation) {}
+    explicit Proxy(const std::string path, bool tor_stream_isolation = false) : m_unix_socket_path(path), m_is_unix_socket(true), m_tor_stream_isolation(tor_stream_isolation) {}
 
     CService proxy;
     std::string m_unix_socket_path;
     bool m_is_unix_socket;
-    bool m_randomize_credentials;
+    bool m_tor_stream_isolation;
```
[the remark this answers is not in the document]
Comment (laanwj) [reply src/netbase.h:68]: CJDNS doesn't use SOCKS, neither does I2P. In any case, randomized credentials leading to circuit isolation is a Tor specific feature. Other uses of SOCKS don't want this (it'd probably either break auth or do nothing), so i still think the new name is better.

File: src/init.cpp (hunk 4 above)
Comment (laanwj) [reply src/init.cpp:1590]: This code is also used by the GUI, not just bitcoind. This is why i say "the application" at the beginning of the sentence instead of being more specific.

File: src/netbase.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (laanwj) [reply src/netbase.h:68]: There's also a deeper conceptual issue that `-proxyrandomize` is a global setting (used for all proxies) instead of a per-proxy setting. In practice no one uses SOCKS for anything else than Tor, let alone multiple different SOCKS proxies at the same time, so it's not something that particularly needs to be addressed.

**Review by hodlinator — APPROVED** [review]
ACK 297ddaa430e99ad62e6bbdcc475ede7e04da547f

#### Concept

Good to avoid Tor circuit sharing for edge-cases like those described in https://github.com/bitcoin/bitcoin/pull/32166#discussion_r2021044241.

Renaming of `randomize_credentials` to `tor_stream_isolation` appears correct. (We support running CJDNS *in conjunction with* Tor SOCKS proxy, https://github.com/bitcoin/bitcoin/pull/32176#discussion_r2024413331).

#### Test

```shell
build/bin/bitcoind -debug=proxy -debug=net -proxy=127.0.0.1:9050
``` 
Log:
```
...
2025-04-02T09:27:27Z [proxy] SOCKS5 sending proxy authentication ec7a807c67cd27ab-0:ec7a807c67cd27ab-0
...
2025-04-02T09:27:47Z [proxy] SOCKS5 sending proxy authentication ec7a807c67cd27ab-1:ec7a807c67cd27ab-1

2025-04-02T09:27:51Z [proxy] Using proxy: 127.0.0.1:9050 to connect to <redacted>:8333
2025-04-02T09:27:51Z [net] SOCKS5 connecting <redacted>
2025-04-02T09:27:51Z [proxy] SOCKS5 sending proxy authentication ec7a807c67cd27ab-2:ec7a807c67cd27ab-2
...
2025-04-02T09:27:58Z [net] SOCKS5 connected <redacted>
2025-04-02T09:27:58Z [net] Added connection peer=1
2025-04-02T09:27:58Z [net] sending version (103 bytes) peer=1
2025-04-02T09:27:58Z [net] send version message: version 70016, blocks=880756, txrelay=1, peer=1
2025-04-02T09:27:58Z [net] start sending v2 handshake to peer=1
2025-04-02T09:27:58Z [net] received: version (102 bytes) peer=1
2025-04-02T09:27:58Z [net] sending wtxidrelay (0 bytes) peer=1
2025-04-02T09:27:58Z [net] sending sendaddrv2 (0 bytes) peer=1
2025-04-02T09:27:58Z [net] sending verack (0 bytes) peer=1
2025-04-02T09:27:58Z [net] sending getaddr (0 bytes) peer=1
2025-04-02T09:27:58Z [net] receive version message: /Satoshi:28.0.0/: version 70016, blocks=890528, us=<redacted>:42840, txrelay=1, peer=1
```

File: src/netbase.cpp [hunk 6]
```diff
@@ -725,6 +725,41 @@ bool IsProxy(const CNetAddr &addr) {
...
+
+    /** Generate a random prefix for each of the credentials returned by this generator.
+     * This makes sure that different launches of the application will not share the same circuits.
+     */
+    static std::string GenerateUniquePrefix() {
+        std::array<uint8_t, PREFIX_BYTE_LENGTH> prefix_bytes;
+        GetRandBytes(prefix_bytes);
+        return HexStr(prefix_bytes) + "-";
+    }
+};
```
[the remark this answers is not in the document]
Comment (laanwj) [reply src/netbase.cpp:761]: tbf i prefer a class, i find that much more readable than a magic function.

**Comment by laanwj** [comment]
[297ddaa430e99ad62e6bbdcc475ede7e04da547f  → ec81a72b369ab9efe23681ebb6e8fab34ce2e0f2](https://github.com/bitcoin/bitcoin/compare/297ddaa430e99ad62e6bbdcc475ede7e04da547f..ec81a72b369ab9efe23681ebb6e8fab34ce2e0f2)
- Updated for @hodlinator's comments

**Comment by laanwj** [comment]
Temporarily cherry-picked #32203 into here for testing. Will revert this branch back to the old commit afterwards.

File: src/netbase.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (hodlinator) [reply src/netbase.cpp]: I admit that the lambda does require a double-take. Maybe I'm having a touch of unexpected C nostalgia.

**Review by hodlinator — APPROVED** [review]
re-ACK ec81a72b369ab9efe23681ebb6e8fab34ce2e0f2

Changes since [first/previous ACK](https://github.com/bitcoin/bitcoin/pull/32176#pullrequestreview-2735670609):
* C++ style closer to doc/developer-notes.md (fixed `class`, still has braces on same line in methods).
* Improved comment about edge-cases fixed by generating unique prefix, and fixed typo.
* Added missing named parameter comments in i2p tests.

File: src/test/fuzz/i2p.cpp [hunk 7]
```diff
@@ -34,7 +34,7 @@ FUZZ_TARGET(i2p, .init = initialize_i2p)
 
     const fs::path private_key_path = gArgs.GetDataDirNet() / "fuzzed_i2p_private_key";
     const CService addr{in6_addr(IN6ADDR_LOOPBACK_INIT), 7656};
-    const Proxy sam_proxy{addr, false};
+    const Proxy sam_proxy{addr, /*tor_stream_isolation=*/false};
```
[the remark this answers is not in the document]
Comment (laanwj) [reply src/test/fuzz/i2p.cpp:37]: Same! Boolean parameters are the least self-descriptive. While making this change i also realized that this field is completely pointless for SAM proxies, which aren't SOCKS, nor have a concept of stream isolation. But using an interchangeable `Proxy` type makes sense to keep `SetProxy` logic simpler, and it's not worth defining a whole type hierarchy for a simple data record.

**Review by jonatack — COMMENTED** [review]
ACK ec81a72b369ab9efe23681ebb6e8fab34ce2e0f2

Adds a randomized prefix as advertised, re-generated on node startup, with an incrementing counter as before for each peer connection. Tested with all p2p networks on (clearnet, tor, i2p, cjdns).

before

```
2025-04-07T18:36:00.698770Z [opencon] [netbase.cpp:415] [Socks5] [proxy] SOCKS5 sending proxy authentication 0:0
...
2025-04-07T18:36:05.967300Z [opencon] [netbase.cpp:415] [Socks5] [proxy] SOCKS5 sending proxy authentication 1:1
```

after

```
2025-04-07T18:34:05.429803Z [opencon] [netbase.cpp:415] [Socks5] [proxy] SOCKS5 sending proxy authentication d1abb33709dc223a-0:d1abb33709dc223a-0
...
2025-04-07T18:34:15.577895Z [opencon] [netbase.cpp:415] [Socks5] [proxy] SOCKS5 sending proxy authentication d1abb33709dc223a-1:d1abb33709dc223a-1
```
restart node, new prefix
```
2025-04-07T18:40:31.064061Z [opencon] [netbase.cpp:415] [Socks5] [proxy] SOCKS5 sending proxy authentication d0b6aa15cd1b1757-0:d0b6aa15cd1b1757-0
...
2025-04-07T18:40:49.239803Z [opencon] [netbase.cpp:415] [Socks5] [proxy] SOCKS5 sending proxy authentication d0b6aa15cd1b1757-1:d0b6aa15cd1b1757-1
...
2025-04-07T18:41:05.524990Z [opencon] [netbase.cpp:415] [Socks5] [proxy] SOCKS5 sending proxy authentication d0b6aa15cd1b1757-2:d0b6aa15cd1b1757-2
...
2025-04-07T18:41:18.697931Z [opencon] [netbase.cpp:415] [Socks5] [proxy] SOCKS5 sending proxy authentication d0b6aa15cd1b1757-3:d0b6aa15cd1b1757-3
...
2025-04-07T18:41:47.225044Z [opencon] [netbase.cpp:415] [Socks5] [proxy] SOCKS5 sending proxy authentication d0b6aa15cd1b1757-4:d0b6aa15cd1b1757-4
```

another restart, new prefix

```
2025-04-07T18:44:18.570628Z [opencon] [netbase.cpp:415] [Socks5] [proxy] SOCKS5 sending proxy authentication 8012260f6062176e-0:8012260f6062176e-0
```

**Review by danielabrozzoni — COMMENTED** [review]
tACK ec81a72b369ab9efe23681ebb6e8fab34ce2e0f2
The code looks good to me.
The way I understand it, Bitcoin Core uses different credentials for each request in order to enable the legacy tor stream isolation, thus using different circuits for each request. This PR introduces a randomized prefix to the credentials that is unique to each session, preventing multiple instances of bitcoin core from reusing the same circuits.

I tested manually and verified that the circuit credentials vary at each run:

```
2025-04-09T17:53:44Z [proxy] SOCKS5 sending proxy authentication fad3434ea2af7c7b-0:fad3434ea2af7c7b-3
[...]
2025-04-09T17:53:47Z [proxy] SOCKS5 sending proxy authentication fad3434ea2af7c7b-1:fad3434ea2af7c7b-4
```
After restarting:

```
2025-04-09T17:53:55Z [proxy] SOCKS5 sending proxy authentication 8ffd3299d2634f06-0:8ffd3299d2634f06-0
```

File: src/netbase.h (hunk 5 above)
Comment (ryanofsky) [inline src/netbase.h:68]: In commit "net: Rename `_randomize_credentials` Proxy parameter to `tor_stream_isolation`" (c47f81e8ac110b1d7f78bce0232e8015366d13e7)

Note: There seems to be a reference to m_randomize_credentials still in a comment https://github.com/bitcoin/bitcoin/blob/51166559808c3528f148b5c0c38cb4481e536dd8/src/torcontrol.cpp#L407

Comment (laanwj) [reply src/netbase.h:68]: Looks like i missed that during rebase as it was introduced in #32166.
